pyccel/ast/cupyext.py

```diff
@@ -0,0 +1,488 @@
+        else:
+            self._start = start
+            self._stop = stop
+        self._step = step if step is not None else LiteralInteger(1)
+
+        if dtype is None:
+            self._dtype = max([i.dtype for i in self.arg], key = NativeNumeric.index)
+            self._precision = max_precision(self.arg, allow_native=False)
+        else:
+            self._dtype, self._precision = process_dtype(dtype)
+
+        self._shape = (MathCeil(PyccelDiv(PyccelMinus(self._stop, self._start), self._step)))
+        self._shape = process_shape(False, self._shape)
+        super().__init__()
+
+    @property
+    def arg(self):
+        return (self._start, self._stop, self._step)
+
+    @property
+    def start(self):
+        return self._start
+
+    @property
+    def stop(self):
+        return self._stop
+
+    @property
+    def step(self):
+        return self._step
+
+#==============================================================================
+
+class Shape(PyccelInternalFunction):
+    """ Represents a call to cupy.shape for code generation
+    """
+    __slots__ = ()
+    name = 'shape'
+    def __new__(cls, arg):
+        if isinstance(arg.shape, PythonTuple):
+            return arg.shape
+        else:
+            return PythonTuple(*arg.shape)
+
+#==============================================================================
+class CupyFull(CupyNewArray):
+    """
+    Represents a call to cupy.full for code generation.
+
+    Parameters
+    ----------
+    shape : int or sequence of ints
+        Shape of the new array, e.g., ``(2, 3)`` or ``2``.
+
+    fill_value : scalar
+        Fill value.
+
+    dtype: str, DataType
+        datatype for the constructed array
+        The default, `None`, means `np.array(fill_value).dtype`.
+
+    order : {'C', 'F'}, optional
+        Whether to store multidimensional data in C- or Fortran-contiguous
+        (row- or column-wise) order in memory.
+
+    """
+    __slots__ = ('_fill_value','_dtype','_precision','_shape','_rank','_order')
+    name = 'full'
+
+    def __init__(self, shape, fill_value, dtype=None, order='C'):
+
+        # Convert shape to PythonTuple
+        shape = process_shape(False, shape)
+        # If there is no dtype, extract it from fill_value
+        # TODO: must get dtype from an annotated node
+        if dtype is None:
+            dtype = fill_value.dtype
+            precision = get_final_precision(fill_value)
+        else:
+            dtype, precision = process_dtype(dtype)
+
+        # Cast fill_value to correct type
+        if fill_value:
+            if fill_value.dtype != dtype or get_final_precision(fill_value) != precision:
+                cast_func = DtypePrecisionToCastFunction[dtype.name][precision]
+                fill_value = cast_func(fill_value)
+        self._shape = shape
+        self._rank  = len(self._shape)
+        self._dtype = dtype
+        self._order = CupyNewArray._process_order(self._rank, order)
```

Comment:
Does `CupyNewArray` need to be named explicitly here? I'm not sure of best practice here. Maybe this is ok, or maybe we should prefer `self` for ease in case of renaming/reorganisation of the inheritance